src/vs/workbench/contrib/scopeTree/browser/bookmarksView.ts

```diff
@@ -20,11 +20,145 @@ import { IBookmarksManager, BookmarkType } from 'vs/workbench/contrib/scopeTree/
 import { Codicon } from 'vs/base/common/codicons';
 import { dirname, basename } from 'vs/base/common/resources';
 import { IExplorerService } from 'vs/workbench/contrib/files/common/files';
+import { IListVirtualDelegate, IKeyboardNavigationLabelProvider } from 'vs/base/browser/ui/list/list';
+import { ScrollbarVisibility } from 'vs/base/common/scrollable';
+import { IDisposable, Disposable } from 'vs/base/common/lifecycle';
+import { WorkbenchObjectTree } from 'vs/platform/list/browser/listService';
+import { FuzzyScore, createMatches } from 'vs/base/common/filters';
+import { ITreeRenderer, ITreeNode, ITreeElement } from 'vs/base/browser/ui/tree/tree';
+import { IResourceLabel, ResourceLabels } from 'vs/workbench/browser/labels';
+
+class Bookmark {
+	public resource: URI;
+
+	constructor(path: string) {
+		this.resource = URI.parse(path);
+	}
+
+	public getName(): string {
+		return basename(this.resource);
+	}
+
+	public getParent(): string {
+		return dirname(this.resource).toString();
+	}
+}
+
+export class BookmarksDelegate implements IListVirtualDelegate<Bookmark> {
+	static readonly ITEM_HEIGHT = 22;
+
+	getHeight(element: Bookmark): number {
+		return BookmarksDelegate.ITEM_HEIGHT;
+	}
+
+	getTemplateId(element: Bookmark): string {
+		return 'BookmarksRenderer';
+	}
+}
+
+interface IBookmarkTemplateData {
+	bookmarkContainer: HTMLElement;
+	label: IResourceLabel;
+	elementDisposable: IDisposable;
+}
+
+class BookmarkDisposable implements IDisposable {
+	private _focusIcon!: HTMLElement;
+
+	constructor(container: HTMLElement,
+		private readonly stat: URI,
+		@IExplorerService private readonly explorerService: IExplorerService) {
+		this.renderFocusIcon(container);
+		this.addListeners(container);
+	}
+
+	get focusIcon(): HTMLElement {
+		return this._focusIcon;
+	}
+
+	private addListeners(container: HTMLElement): void {
+		container.addEventListener('mouseover', () => {
+			this._focusIcon.style.visibility = 'visible';
+		});
+		container.addEventListener('mouseout', () => {
+			this._focusIcon.style.visibility = 'hidden';
+		});
+		container.addEventListener('dblclick', async () => {
+			await this.explorerService.select(this.stat, true);	// Should also expand directory
+		});
+		this._focusIcon.addEventListener('click', () => {
+			this.explorerService.setRoot(this.stat);
+		});
+	}
+
+	private renderFocusIcon(container: HTMLElement): void {
+		this._focusIcon = document.createElement('img');
+		this._focusIcon.className = 'scope-tree-focus-icon-near-bookmark';
+		container.insertBefore(this._focusIcon, container.firstChild);
+	}
+
+	dispose(): void {
+		this._focusIcon.remove();
+	}
+}
+
+class BookmarksRenderer implements ITreeRenderer<Bookmark, FuzzyScore, IBookmarkTemplateData> {
+	static readonly ID = 'BookmarksRenderer';
+
+	constructor(
+		private labels: ResourceLabels,
+		private readonly explorerService: IExplorerService
+	) {
+		// noop
+	}
```

Comment:
In that case you just have it like this:

```suggestion
	constructor(
		private labels: ResourceLabels,
		private readonly explorerService: IExplorerService
	) {}
```